File: Techniques_Code Optimization/Exercises/greatest_product.rb
def greatest_product(array)
    greatest_number = -Float::INFINITY 
    second_to_greatest_number = -Float::INFINITY
    
    lowest_number = Float::INFINITY
    second_to_lowest_number = Float::INFINITY    
   
    array.each do |number|
        if number >= greatest_number
            second_to_greatest_number = greatest_number
            greatest_number = number
        elsif number > second_to_greatest_number
            second_to_greatest_number = number
        end
    
        if number <= lowest_number 
            second_to_lowest_number = lowest_number 
            lowest_number = number
        elsif number > lowest_number && number < second_to_lowest_number 
            second_to_lowest_number = number
        end 
    end
    
    greatest_product_from_two_highest = greatest_number * second_to_greatest_number
    greatest_product_from_two_lowest = lowest_number * second_to_lowest_number
    
    if greatest_product_from_two_highest > greatest_product_from_two_lowest 
        return greatest_product_from_two_highest
    else
        return greatest_product_from_two_lowest
    end 
end

puts greatest_product([3, -10, -6, 9, 4])

File: Techniques_Code Optimization/Exercises/greatestProduct.js
// Time complexity: O(n)
// Space complexity: O(1)
const greatestProduct = (array) => {
  let greatestNumber = -Infinity;
  let secondToGreatestNumber = -Infinity;

  let lowestNumber = Infinity;
  let secondToLowestNumber = Infinity;

  // Greedy approach
  for (const number of array) {
    if (number >= greatestNumber) {
      secondToGreatestNumber = greatestNumber;
      greatestNumber = number;
    } else if (number > secondToGreatestNumber) {
      secondToGreatestNumber = number;
    }

    if (number <= lowestNumber) {
      secondToLowestNumber = lowestNumber;
      lowestNumber = number;
    } else if (number > lowestNumber && number < secondToLowestNumber) {
      secondToLowestNumber = number;
    }
  }

  // Pattern recognized from generating examples
  const greatestProductFromTwoHighest = greatestNumber * secondToGreatestNumber;
  const greatestProductFromTwoLowest = lowestNumber * secondToLowestNumber;

  if (greatestProductFromTwoHighest > greatestProductFromTwoLowest) {
    return greatestProductFromTwoHighest;
  } else {
    return greatestProductFromTwoLowest;
  }
};

// Examples
greatestProduct([3, -10, -6, 9, 4]); // should return 60
greatestProduct([2, 6 - 4, 9, 10, 11]); // should return 110
greatestProduct([-3, 10, 1]); // should return 10
greatestProduct([-3, 2, 10, 1]); // should return 20
greatestProduct([-3, 2, -10, 1]); // should return 30
greatestProduct([1, -3]); // should return -3
